### ISSUE TITLE
Terminated not killing sinks

### ISSUE THREAD

**Issue by djspiewak** [scalaz/scalaz-stream#402]
Would we expect these tests to pass?  This is against 0.7.1.

```scala
  property("to.halt") = secure {
    var count = 0

    val src = Process.range(0, 10) evalMap { i =>
      Task delay {
        count += 1
        i
      }
    }

    val ch = sink lift { _: Int => (Task delay { throw Terminated(End); () }) }

    val results = (src observe ch runLog).run

    results.isEmpty :| s"expected empty; got $results" &&
      (count === 0) :| s"count = $count"
  }

  property("observe.halt") = secure {
    val src = Process.range(0, 10)
    val ch = sink lift { _: Int => (Task delay { throw Terminated(End); () }) }

    val results = (src observe ch runLog).run
    results.isEmpty :| s"expected empty; got $results"
  }
```

Interestingly, they also fail (but fail *differently*) in master.

**Comment by pchlupacek** [comment]
I think `toHalt` shall have count eq 1 , as it is evaluated before going to tee. `results` imho shall be empty, so that prop shall pass. So this shall fail on `count` prop. 

`observe.halt` shall pass I guess.

**Comment by djspiewak** [comment]
I think you're right about count, but neither test passes right now. They all go up to 10.

**Comment by pchlupacek** [comment]
:-) I suspect we have review really this terminated wizardry :-) In our code we usually use acquire.flatMap(constantSink(ofResource)) kind of sinks, so this is something we didn't really test. 

LMK if you will find something about the reason.

**Comment by djspiewak** [comment]
@pchlupacek I think I figured it out.  The following is a more minimized case which exhibits the same bug:

```scala
var count = 0

val src = Process.eval(Task delay {
  count += 1
}) ++ Process.eval(Task delay {
  count += 1
})

(src flatMap { _ => Process.eval(Task delay { throw Terminated(End); () }) }).run.run
```

After running the above, `count` will be `2`.  However, the following example is different:

```scala
var count = 0

val src = Process.eval(Task delay {
  count += 1
}) ++ Process.eval(Task delay {
  count += 1
})

(src flatMap { _ => Process.eval(Task delay { throw Terminated(Kill); () }) }).run.run
```

In this case, `count` is `1`.  Of course, it's clear what is happening here.

The problem is that `Append` is blithely eating the `End` that we get from transmuting the `Terminated` and treating it as a signal to "move on to the next thing".  `repeatEval` doesn't have this problem, of course, but we have numerous cases (like `observe` and `through`) where we are `eval`ing individual actions within an infinitely appended stream.

In master, `repeatEval` is no different from `eval . repeat`, and also the automagical transmutation of `Terminated` no longer happens.  So the semantics are different, and in fact *closer* to what one would logically expect from throwing `Terminated(End)` within an effect inside of a channel.  However, the differences appear to be enough to cause other…issues.

Anyway, the tl;dr is that things are very very broken, and I'm concerned that we (by which I mean, scalaz-stream) have code which is *relying* on this broken behavior.

**Comment by pchlupacek** [comment]
@djspiewak good observation. I think we know the Teriminated is unfortuantelly sort of hack. Any reason why `automagic` can't be put back at least for `eval` before we will figure out how to do this correcly?

**Comment by djspiewak** [comment]
@pchlupacek I think we might need to put it back.  The crazy thing is that we're just restoring behavior that's broken.  The behavior in master is *less* broken and inscrutable than the behavior in 0.7, it's just that the behavior in 0.7 is familiar (or at least has familiar consequences).

**Comment by pchlupacek** [comment]
:-) unfortunately. Is as always: Easiness and correctness is fairly complicated to do together :-) I think we likely to address this with 0.8.

**Comment by djspiewak** [comment]
@pchlupacek I was able to find a solution that works in master.  I can't fix series/0.7 without *significant* shenanigans, not to mention silently breaking compatibility (albeit compatibility with a bugged semantic).    The long-term solution to fixing all of our `Terminated` woes is really to achieve some sort of transducers encoding, so that the termination state (which is a special case, but still important) is not such a hack in channels and sinks.

**Comment by pchlupacek** [comment]
@djspiewak cool and thanks. I think it is still beneficial to have master in good shape, even when perhaps we won't do any binary compatible branch of it anymore. Hence we still don't have that algebra fixed, there is still chance that we do interim release. 

Thanks!
